fix(client): type inference for `prismaClient.$on(...)`

nikelborm opened the pull request:
# Problem

1. incorrect inference in generic `GetLogType` returned `never` instead of `T` on line 6. ([original piece of code here](https://github.com/prisma/prisma/blob/main/packages/client/src/generation/TSClient/PrismaClient.ts#L555))
```typescript
/*1.*/export type GetLogType<T extends LogLevel | LogDefinition> =
/*2.*/  T extends LogDefinition
/*3.*/    ? T['emit'] extends 'event'
/*4.*/      ? T['level']
/*5.*/      : never
/*6.*/  : never
```
This code led to error when some of the elements in `T` argument of `GetEvents` which are being later passed to `GetLogType` were `LogLevel` string literals instead of `LogDefinition` objects.

2. Case when developer forgot to add `as const` (and there's no suggestion in documentation to add it) to the argument of `PrismaClient` constructor.

[Example in documentation:](https://github.com/prisma/docs/blob/main/content/200-orm/200-prisma-client/600-observability-and-logging/130-logging.mdx?plain=1#L34-L37)
```typescript
const prisma = new PrismaClient({
  log: ['query', 'info', 'warn', 'error'],
})
```

Since neither developer, nor `PrismaClient` generic itself (`const` before `U` is absent [here](https://github.com/prisma/prisma/blob/main/packages/client/src/generation/TSClient/PrismaClient.ts#L467)) didn't tell typescript to treat argument of `new PrismaClient(...)` as constant, the type information of string literals will be erased and typescript will handle it just as `{ log: string[] }`.

3. `GetEvents` was hardcoded to expect arrays with length at max 4-elements [here](https://github.com/prisma/prisma/blob/main/packages/client/src/generation/TSClient/PrismaClient.ts#L557)

# Solution

1. `GetLogType` now returns those literals as is when they don't extend `LogDefinition` (object), but also with additional check which then forces `T` of `GetLogType<T>` to extend `LogLevel`
2. Added useful `const` modifier to the second `U` argument of `PrismaClient` generic which preserves string literal type info taken from the argument of `PrismaClient` constructor
3. `GetEvents` generic is now a ~~recursive~~ **UPD distributed** type and doesn't depend on length of `PrismaClientOptions.log` array

Might fix some of the following issues:
1. https://github.com/prisma/prisma/issues/19463
2. https://github.com/prisma/docs/issues/5021
3. https://github.com/prisma/docs/issues/1564
4. https://github.com/prisma/prisma/issues/4746
5. https://github.com/prisma/prisma/issues/7068
6. https://github.com/notiz-dev/nestjs-prisma/issues/56
7. https://github.com/prisma/prisma/issues/11986
8. https://github.com/prisma/prisma/issues/6279
9. https://github.com/prisma/docs/issues/3424

## Before
`U` argument of `PrismaClient` generic is `never` mainly because of the 1st problem mentioned above and we can't call `prismaClient.$on(...)` passing any string literal to it:
![Screenshot from 2024-03-05 14-23-28](https://github.com/prisma/prisma/assets/29357907/e52c397e-6e68-4d1f-aa57-6795c1c07044)

## After
The 1st screenshot shows that the second inferred argument `U` of generic `PrismaClient` stored in `DB` constant is now `"info" | "query" | "error"` according to IDE's tooltip above the line `DB.$on('', () => {})` and typescript also tells to replace first argument `''` to any of `U` union members: 
![Screenshot from 2024-03-05 14-37-44](https://github.com/prisma/prisma/assets/29357907/96822786-b04e-4914-a410-9f32a6673319)

2nd screenshot shows that when we pass correct 1st argument to `DB.$on` it correctly infers type of the 2nd function argument:
![Screenshot from 2024-03-05 14-41-01](https://github.com/prisma/prisma/assets/29357907/49ea86b0-5c11-4261-8b68-f43a036bec30)

<!-- num: 3-->
<!-- note to me: https://github.com/sandersn/downlevel-dts -->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/prisma/prisma?pullRequest=24133) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/prisma/prisma?pullRequest=24133) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/prisma/prisma?pullRequest=24133) it.</sub>

kduprey commented:
Any updates or progress on this? Are we able to get this merged? 
Running into this issue when attempting to customise the `prismaClient` event logging.

aqrln commented on packages/client/src/generation/TSClient/PrismaClient.ts:
```diff
@@ -464,7 +464,7 @@ export class PrismaClientClass implements Generable {
 export class PrismaClient<
   ClientOptions extends Prisma.PrismaClientOptions = Prisma.PrismaClientOptions,
-  U = 'log' extends keyof ClientOptions ? ClientOptions['log'] extends Array<Prisma.LogLevel | Prisma.LogDefinition> ? Prisma.GetEvents<ClientOptions['log']> : never : never,
+  const U = 'log' extends keyof ClientOptions ? ClientOptions['log'] extends Array<Prisma.LogLevel | Prisma.LogDefinition> ? Prisma.GetEvents<ClientOptions['log']> : never : never,
```
This is a breaking change. Const generics require TypeScript 5.0 or later but Prisma supports 4.7+ right now. We're raising the minimal required version to 5.1 soon in Prisma 6 though.

nikelborm replied:
Yeah... Unfortunately it (`Const generics require TypeScript 5.0 or later`) does...
But you can cherry pick one commit and not touch the other one.
Or I can create 2 split PRs. One of them for the closest version and the second one is for 6.0
What do you think?

aqrln replied:
Prisma 6 is around the corner so let's just merge the whole PR.

aqrln reviewed (commented):
Thanks for the PR and sorry it was overlooked! While this can't land in Prisma 5.x (unless the solution can be changed to not rely on TypeScript 5 features), this should definitely be merged for upcoming Prisma 6.

(cc @jkomyno)

aqrln commented:
@nikelborm could you please rebase your PR onto latest `main` so you don't have TS 4.7 failures (we already bumped the minimal required TypeScript version on main)? There are also other things failing in https://github.com/prisma/prisma/actions/runs/11617514866/job/32352930786?pr=24133 that need to be looked into:

```
  ● valid types › /home/runner/work/prisma/prisma/packages/client/src/__tests__/types/deprecation

    expect(received).resolves.not.toThrow()

    Received promise rejected instead of resolved
    Rejected to value: [Error: Command failed with exit code 1: /opt/hostedtoolcache/node/16.20.2/x64/bin/node /home/runner/work/prisma/prisma/packages/client/src/utils/compilerWorker.js /home/runner/work/prisma/prisma/packages/client/src/__tests__/types/deprecation/test.ts
    /home/runner/work/prisma/prisma/packages/client/src/utils/compilerWorker.js:20
        throw new Error('Compilation Error\n' + formatted)
        ^·
    Error: Compilation Error
    src/__tests__/types/deprecation/test.ts(20,14): error TS2345: Argument of type 'string' is not assignable to parameter of type 'never'.·
```

nikelborm commented:
> @nikelborm could you please rebase your PR onto latest `main` so you don't have TS 4.7 failures (we already bumped the minimal required TypeScript version on main)? There are also other things failing in [prisma/prisma/actions/runs/11617514866/job/32352930786?pr=24133](https://github.com/prisma/prisma/actions/runs/11617514866/job/32352930786?pr=24133) that need to be looked into:
> 
> ```
>   ● valid types › /home/runner/work/prisma/prisma/packages/client/src/__tests__/types/deprecation
> 
>     expect(received).resolves.not.toThrow()
> 
>     Received promise rejected instead of resolved
>     Rejected to value: [Error: Command failed with exit code 1: /opt/hostedtoolcache/node/16.20.2/x64/bin/node /home/runner/work/prisma/prisma/packages/client/src/utils/compilerWorker.js /home/runner/work/prisma/prisma/packages/client/src/__tests__/types/deprecation/test.ts
>     /home/runner/work/prisma/prisma/packages/client/src/utils/compilerWorker.js:20
>         throw new Error('Compilation Error\n' + formatted)
>         ^·
>     Error: Compilation Error
>     src/__tests__/types/deprecation/test.ts(20,14): error TS2345: Argument of type 'string' is not assignable to parameter of type 'never'.·
> ```

Rebased and fixed the issue. Not sure about those 2 failing CI's left. They seem to be failing not because of my changes

aqrln commented on packages/client/src/__tests__/integration/happy/not-so-exhaustive-schema-mongo/__snapshots__/binary.test.ts.snap:
```diff
@@ -2277,10 +2277,15 @@ export namespace Prisma {
+
+  export type GetLogType<T> = T extends LogDefinition
+    ? (T['emit'] extends 'event' ? CheckIsLogLevel<T['level']> : never)
+    : CheckIsLogLevel<T>;
```
So, about this. String literals are a shorthand for `{ level: "...", emit: "stdout" }`, so it was intentional that `never` was returned in this case, since [according to the docs](https://www.prisma.io/docs/orm/prisma-client/observability-and-logging/logging) it's either events or logs in stdout but not both.

However, this is not the actual behavior. You will indeed receive both events and console logs when using either the shorthand or the full `emit: "stdout"` form.

Technically this is a bug, as it contradicts both the types and the documentation. However, since we're doing a major release anyway, I'm fine with changing the semantics the way you propose and updating the types and the docs to match the actual behavior because it makes more sense to me too. In that case, though, we shouldn't restrict the `LogDefinition` case to `T['emit'] extends 'event'` either.

nikelborm replied:
Do you mean changing it from `T['emit'] extends 'event' ? CheckIsLogLevel<T['level']> : never` to just `CheckIsLogLevel<T['level']>`, right?

Also did I understand correctly that you want me to change the docs too?

aqrln replied:
> Do you mean changing it from T['emit'] extends 'event' ? CheckIsLogLevel<T['level']> : never to just CheckIsLogLevel<T['level']>, right?

Yes

> Also did I understand correctly that you want me to change the docs too? I'm happy to do it

That would be amazing, but up to you, we can also do this. Probably a small note that tells that `emit: 'stdout'` logs are also emitted as events is sufficient.

nikelborm committed:
fix(client): added const to generic for better type inference

nikelborm committed:
fix(client): type inference for prismaClient.$on(...)

aqrln committed:
Merge branch 'main' into main

aqrln committed:
Merge remote-tracking branch 'origin/main' into nikelborm/main

aqrln committed:
Port changes to the new generator and update doc comments

aqrln reviewed (approved):
LGTM, sorry for this falling through the cracks — don't hesitate to ping the reviewers in the future. I've updated the PR to port the changes to the new generator and updated the documentation comments.

aqrln committed:
Remove invalid const generic on interface
